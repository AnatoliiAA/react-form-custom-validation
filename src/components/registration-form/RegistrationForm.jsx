import { useState } from "react";
import { Input } from "../form-inputs/input/Input";
import { PasswordInput } from "../form-inputs/password-input/PasswordInput";
import { Checkbox } from "../form-inputs/checkbox-input/Checkbox";
import { validateField, checkValidationObj } from "../../helpers";
import "./RegistrationForm.scss";

export const RegistrationForm = () => {
  const [formValues, setFormValues] = useState({
    firstName: "",
    lastName: "",
    phone: "",
    email: "",
    password: "",
    repeatPassword: "",
    acceptReceiveEmails: false,
    acceptTerms: false,
  });

  const [inputsValid, setInputsValid] = useState({
    firstName: false,
    lastName: false,
    phone: false,
    email: false,
    password: false,
    repeatPassword: false,
    acceptTerms: false,
  });

  const [showError, setShowError] = useState({
    firstName: false,
    lastName: false,
    phone: false,
    email: false,
    password: false,
    repeatPassword: false,
    acceptTerms: false,
  });

  const formErrors = {
    firstName:
      "First Name should be at least 2 characters long and contain only alphabetic",
    lastName:
      "Last Name should be at least 2 characters long and contain only alphabetic",
    phone:
      "Phone should contain 12 digits and can have foollowing format: '+xxxxxxxxxxxx' or 'xxxxxxxxxxxx'",
    email: "You should enter a valid email",
    password: "Password should be at least 8 characters long",
    repeatPassword: "Passwords should match",
    acceptTerms: "This field is required",
  };

  const formClassName = "registration-form";

  const termsLinks = (
    <p>
      <a href="#" className={`${formClassName}__link`}>
        &nbsp;Term,{" "}
      </a>
      <a href="#" className={`${formClassName}__link`}>
        Privacy Policy{" "}
      </a>
      and{" "}
      <a href="#" className={`${formClassName}__link`}>
        Fees
      </a>
    </p>
  );

  const inputsData = [
    {
      type: "text",
      name: "firstName",
      id: "first-name",
      labelText: "First Name",
    },
    { type: "text", name: "lastName", id: "last-name", labelText: "Last Name" },
    {
      type: "tel",
      name: "phone",
      id: "phone-number",
      labelText: "Phone Number",
    },
    { type: "email", name: "email", id: "email", labelText: "Email" },
    {
      type: "password",
      name: "password",
      id: "password",
      labelText: "Password",
    },
    {
      type: "password",
      name: "repeatPassword",
      id: "repeat-password",
      labelText: "Repeat Password",
    },
  ];

  const createInput = (type, name, attrs) =>
    type === "password" ? (
      <PasswordInput
        name={name}
        onChange={handleInputChange}
        value={formValues[name]}
        validity={inputsValid[name]}
        error={formErrors[name]}
        showError={showError[name]}
        {...attrs}
      />
    ) : type === "checkbox" ? (
      <Checkbox
        name={name}
        onChange={handleInputChange}
        checked={formValues[name]}
        validity={inputsValid[name]}
        error={formErrors[name]}
        showError={showError[name]}
        {...attrs}
      />
    ) : (
      <Input
        type={type}
        name={name}
        onChange={handleInputChange}
        value={formValues[name]}
        validity={inputsValid[name]}
        error={formErrors[name]}
        showError={showError[name]}
        {...attrs}
      />
    );

  const handleInputChange = (e) => {
    const value =
      e.target.type === "checkbox" ? e.target.checked : e.target.value;
    const name = e.target.name;

    const setValidityAndError = (name, validity) => {
      setInputsValid({ ...inputsValid, [name]: validity });
      setShowError({ ...showError, [name]: !validity });
    };

    const validateValue = (name, value) => {
      let validity;
      switch (name) {
        case "repeatPassword":
          validity = value === formValues.password ? true : false;
          break;
        case "acceptTerms":
          validity = !formValues.acceptTerms;
          break;
        default:
          validity = validateField(name, value);
      }
      return validity;
    };

    const validity = validateValue(name, value);
    setValidityAndError(name, validity);

    if (value === "") {
      setShowError({ ...showError, [name]: false });
    }

    setFormValues({ ...formValues, [name]: value });
  };

  const sendFormData = (dataObj) => {
    const response = fetch("https://google.com/", {
      method: "POST",
      mode: "no-cors",
      cache: "no-cache",
      credentials: "same-origin",
      headers: {
        "Content-Type": "application/json",
      },
      redirect: "follow",
      referrerPolicy: "no-referrer",
      body: JSON.stringify(dataObj),
    })
      .then((result) => {
        console.log(result);
      })
      .catch((error) => {
        console.log(error);
      });
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    const formIsValid = checkValidationObj(inputsValid);
    if (formIsValid) {
      sendFormData(formValues);
      setFormValues({
        firstName: "",
        lastName: "",
        phone: "",
        email: "",
        password: "",
        repeatPassword: "",
        acceptReceiveEmails: false,
        acceptTerms: false,
      });
      setInputsValid({
        firstName: false,
        lastName: false,
        phone: false,
        email: false,
        password: false,
        repeatPassword: false,
        acceptTerms: false,
      });
      return;
    }

    const newShowError = { ...inputsValid };
    for (let key in newShowError) {
      newShowError[key] = !newShowError[key];
    }
    setShowError(newShowError);
  };

  return (
    <div className="registration-wrapper">
      <h1 className={`${formClassName}__title`}>Register</h1>
      <h2 className={`${formClassName}__header`}>
        Manage all your lottery efficiently
      </h2>
      <p className={`${formClassName}__text`}>
        Let's get you all set up so you can verify your personal account and
        begin setting up your profile
      </p>
      <form className={formClassName} onSubmit={handleSubmit}>
        {inputsData.map((inputData) =>
          createInput(inputData["type"], inputData["name"], {
            id: inputData["id"],
            labelText: inputData["labelText"],
          })
        )}
        <div
          className={`${formClassName}__column ${formClassName}__column--wide`}
        >
          {createInput("checkbox", "acceptReceiveEmails", {
            id: "receive-emails",
            labelText: "Yes, I want to receive Lottery Display emails",
            validity: "",
          })}
          {createInput("checkbox", "acceptTerms", {
            id: "accept-terms",
            labelText: "I agree to the all",
            children: termsLinks,
          })}
        </div>
        <button className={`${formClassName}__submit`} type="submit">
          Create Account
        </button>
        <div
          className={`${formClassName}__column ${formClassName}__column--wide`}
        >
          <p>
            Already have an account?{" "}
            <a href="#" className={`${formClassName}__link`}>
              Log In
            </a>
          </p>
        </div>
      </form>
    </div>
  );
};
