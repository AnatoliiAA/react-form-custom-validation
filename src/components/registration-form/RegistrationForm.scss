@import "../../styles/vars";
@import "../../styles/color-palette";

.registration-wrapper {
  max-width: 800px;
  box-sizing: border-box;
  margin: 0;
  padding: 20px 10px;
  text-align: left;
  color: $color-dark-blue;
}

.registration-form {
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
  font-size: $font-size-normall;
  &__title {
    margin: 0 0 60px 0;
    font-size: $font-size-xlarge;
    font-weight: bold;
  }
  &__header {
    margin: 0 0 20px 0;
    font-size: $font-size-large;
    font-weight: bold;
  }
  &__text {
    margin: 0 0 40px 0;
  }
  &__column {
    &--wide {
      flex-direction: row;
      flex-wrap: wrap;
      width: 100%;
    }
  }
  &__checkbox-wrapper {
    display: flex;
    align-items: center;
    width: 100%;
    height: 30px;
    margin-bottom: 20px;
    & .registration-form__label {
      font-weight: 400;
    }
  }
  &__link {
    text-decoration: none;
    color: $color-light-blue;
    &:hover {
      border-bottom: 1px solid $color-light-blue;
    }
  }
  &__submit {
    width: 250px;
    height: 40px;
    margin: 30px 0;
    font-size: $font-size-normall;
    font-weight: bold;
    color: $color-white;
    background-color: $color-light-blue;
    border: 0px;
    border-radius: 10px;
    cursor: pointer;
    &:hover {
      border: 2px solid $color-dark-blue;
    }
  }
}
