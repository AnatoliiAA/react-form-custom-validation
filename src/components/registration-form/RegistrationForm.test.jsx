import React from "react";
import { render, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom";
import { RegistrationForm } from "./RegistrationForm";

const selectors = {
  name: "#first-name",
  lastName: "#last-name",
  phone: "#phone-number",
  password: "#password",
  repeatPassword: "#repeat-password",
  email: "#email",
  submitButton: ".registration-form__submit",
  inputError: ".text-input__error",
  checkboxError: ".checkbox-input__error",
  acceptTerms: "#accept-terms",
  passwordVisibility: ".password-visibility"
};

describe("registration-form tests", () => {
  describe("render content tests", () => {
    it("should render a component", () => {
      const { container } = render(<RegistrationForm />);
      expect(container).toBeInTheDocument();
    });
    it("should match snapshot", () => {
      const tree = renderer.create(<RegistrationForm />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("actions tests", () => {
    it("should show 7 errors on submit if form is empty", () => {
      const { container } = render(<RegistrationForm />);
      const submitButton = container.querySelector(selectors["submitButton"]);

      fireEvent.click(submitButton);

      const errors = container.querySelectorAll(selectors["inputError"]);
      const checkboxError = container.querySelector(selectors["checkboxError"]);
      expect(errors.length).toBe(6);
      expect(checkboxError).toBeInTheDocument();
    });

    it("should show error on ivalid first name input", () => {
      const { container } = render(<RegistrationForm />);
      const nameInput = container.querySelector(selectors["name"]);

      fireEvent.change(nameInput, { target: { value: "123 Invalid" } });

      const error = container.querySelector(selectors["inputError"]);
      expect(error).toBeInTheDocument();
      expect(error).toHaveTextContent(
        "First Name should be at least 2 characters long and contain only alphabetic"
      );
    });

    it("should show error on ivalid last name input", () => {
      const { container } = render(<RegistrationForm />);
      const nameInput = container.querySelector(selectors["lastName"]);

      fireEvent.change(nameInput, { target: { value: "123 Invalid" } });

      const error = container.querySelector(selectors["inputError"]);
      expect(error).toBeInTheDocument();
      expect(error).toHaveTextContent(
        "Last Name should be at least 2 characters long and contain only alphabetic"
      );
    });

    it("should show error on ivalid phone number input", () => {
      const { container } = render(<RegistrationForm />);
      const phoneInput = container.querySelector(selectors["phone"]);

      fireEvent.change(phoneInput, { target: { value: "invalid" } });

      const error = container.querySelector(selectors["inputError"]);
      expect(error).toBeInTheDocument();
      expect(error).toHaveTextContent(
        "Phone should contain 12 digits and can have foollowing format: '+xxxxxxxxxxxx' or 'xxxxxxxxxxxx'"
      );
    });

    it("should show error on ivalid email input", () => {
      const { container } = render(<RegistrationForm />);
      const emailInput = container.querySelector(selectors["email"]);

      fireEvent.change(emailInput, { target: { value: "invalid" } });

      const error = container.querySelector(selectors["inputError"]);
      expect(error).toBeInTheDocument();
      expect(error).toHaveTextContent("You should enter a valid email");
    });

    it("should show error on ivalid password input", () => {
      const { container } = render(<RegistrationForm />);
      const passwordInput = container.querySelector(selectors["password"]);

      fireEvent.change(passwordInput, { target: { value: "1" } });

      const error = container.querySelector(selectors["inputError"]);
      expect(error).toBeInTheDocument();
      expect(error).toHaveTextContent(
        "Password should be at least 8 characters long"
      );
    });

    it("should show error on ivalid repeat password input", () => {
      const { container } = render(<RegistrationForm />);
      const passwordInput = container.querySelector(selectors["password"]);
      const repeatPasswordInput = container.querySelector(selectors["repeatPassword"]);

      fireEvent.change(passwordInput, { target: { value: "123456789" } });
      fireEvent.change(repeatPasswordInput, {
        target: { value: "987654321" },
      });

      const error = container.querySelector(selectors["inputError"]);
      expect(error).toBeInTheDocument();
      expect(error).toHaveTextContent("Passwords should match");
    });

    it("should show an error if you uncheck required checkbox", () => {
      const { container } = render(<RegistrationForm />);
      const checkboxInput = container.querySelector(selectors["acceptTerms"]);

      fireEvent.click(checkboxInput);
      fireEvent.click(checkboxInput);

      const error = container.querySelector(selectors["checkboxError"]);
      expect(error).toBeInTheDocument();
      expect(error).toHaveTextContent("This field is required");
    });

    it("should not show error after clearing field with invalid data", () => {
      const { container } = render(<RegistrationForm />);
      const nameInput = container.querySelector(selectors["name"]);
      fireEvent.change(nameInput, { target: { value: "123 Invalid" } });
      const error = container.querySelector(selectors["inputError"]);
      fireEvent.change(nameInput, { target: { value: "" } });
      expect(error).not.toBeInTheDocument();
    });

    it("should show password after clicking on visibility button", () => {
      const { container } = render(<RegistrationForm />);
      const passwordInput = container.querySelector(selectors["password"]);
      const visibilityButton = container.querySelector(".password-visibility");

      fireEvent.click(visibilityButton);

      expect(passwordInput.type).toBe("text");
    });

    it("should clear form on submit if every required field is filled correctly", () => {
      const { container } = render(<RegistrationForm />);
      const submitButton = container.querySelector(
        ".registration-form__submit"
      );
      const nameInput = container.querySelector(selectors["name"]);
      const lastnameInput = container.querySelector(selectors["lastName"]);
      const phoneInput = container.querySelector(selectors["phone"]);
      const emailInput = container.querySelector(selectors["email"]);
      const passwordInput = container.querySelector(selectors["password"]);
      const repeatPasswordInput = container.querySelector(selectors["repeatPassword"]);
      const checkboxInput = container.querySelector(selectors["acceptTerms"]);

      fireEvent.change(nameInput, { target: { value: "Valid" } });
      fireEvent.change(lastnameInput, { target: { value: "Valid" } });
      fireEvent.change(phoneInput, { target: { value: "123123123123" } });
      fireEvent.change(emailInput, { target: { value: "email@email.com" } });
      fireEvent.change(passwordInput, { target: { value: "12345678" } });
      fireEvent.change(repeatPasswordInput, {
        target: { value: "12345678" },
      });
      fireEvent.click(checkboxInput);
      fireEvent.click(submitButton);

      expect(nameInput).toHaveValue("");
      expect(lastnameInput).toHaveValue("");
      expect(phoneInput).toHaveValue("");
      expect(emailInput).toHaveValue("");
      expect(passwordInput).toHaveValue("");
      expect(repeatPasswordInput).toHaveValue("");
      expect(checkboxInput).not.toBeChecked();
    });
  });
});
