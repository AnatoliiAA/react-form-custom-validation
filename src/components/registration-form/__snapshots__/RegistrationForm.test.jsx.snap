// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registration-form tests render content tests should match snapshot 1`] = `
<div
  className="registration-wrapper"
>
  <h1
    className="registration-form__title"
  >
    Register
  </h1>
  <h2
    className="registration-form__header"
  >
    Manage all your lottery efficiently
  </h2>
  <p
    className="registration-form__text"
  >
    Let's get you all set up so you can verify your personal account and begin setting up your profile
  </p>
  <form
    className="registration-form"
    onSubmit={[Function]}
  >
    <div
      className="input-wrapper"
    >
      <label
        className="text-input-label"
        htmlFor="first-name"
      >
        First Name
        
      </label>
      <input
        className="text-input text-input--invalid text-input--empty"
        id="first-name"
        name="firstName"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="input-wrapper"
    >
      <label
        className="text-input-label"
        htmlFor="last-name"
      >
        Last Name
        
      </label>
      <input
        className="text-input text-input--invalid text-input--empty"
        id="last-name"
        name="lastName"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="input-wrapper"
    >
      <label
        className="text-input-label"
        htmlFor="phone-number"
      >
        Phone Number
        
      </label>
      <input
        className="text-input text-input--invalid text-input--empty"
        id="phone-number"
        name="phone"
        onChange={[Function]}
        type="tel"
        value=""
      />
    </div>
    <div
      className="input-wrapper"
    >
      <label
        className="text-input-label"
        htmlFor="email"
      >
        Email
        
      </label>
      <input
        className="text-input text-input--invalid text-input--empty"
        id="email"
        name="email"
        onChange={[Function]}
        type="email"
        value=""
      />
    </div>
    <div
      className="input-wrapper"
    >
      <label
        className="text-input-label"
        htmlFor="password"
      >
        Password
        
      </label>
      <input
        className="text-input text-input--invalid text-input--empty"
        id="password"
        name="password"
        onChange={[Function]}
        type="password"
        value=""
      />
      <span
        className="password-visibility"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-eye-slash fa-w-20 "
          data-icon="eye-slash"
          data-prefix="far"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 640 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M634 471L36 3.51A16 16 0 0 0 13.51 6l-10 12.49A16 16 0 0 0 6 41l598 467.49a16 16 0 0 0 22.49-2.49l10-12.49A16 16 0 0 0 634 471zM296.79 146.47l134.79 105.38C429.36 191.91 380.48 144 320 144a112.26 112.26 0 0 0-23.21 2.47zm46.42 219.07L208.42 260.16C210.65 320.09 259.53 368 320 368a113 113 0 0 0 23.21-2.46zM320 112c98.65 0 189.09 55 237.93 144a285.53 285.53 0 0 1-44 60.2l37.74 29.5a333.7 333.7 0 0 0 52.9-75.11 32.35 32.35 0 0 0 0-29.19C550.29 135.59 442.93 64 320 64c-36.7 0-71.71 7-104.63 18.81l46.41 36.29c18.94-4.3 38.34-7.1 58.22-7.1zm0 288c-98.65 0-189.08-55-237.93-144a285.47 285.47 0 0 1 44.05-60.19l-37.74-29.5a333.6 333.6 0 0 0-52.89 75.1 32.35 32.35 0 0 0 0 29.19C89.72 376.41 197.08 448 320 448c36.7 0 71.71-7.05 104.63-18.81l-46.41-36.28C359.28 397.2 339.89 400 320 400z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </span>
    </div>
    <div
      className="input-wrapper"
    >
      <label
        className="text-input-label"
        htmlFor="repeat-password"
      >
        Repeat Password
        
      </label>
      <input
        className="text-input text-input--invalid text-input--empty"
        id="repeat-password"
        name="repeatPassword"
        onChange={[Function]}
        type="password"
        value=""
      />
      <span
        className="password-visibility"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-eye-slash fa-w-20 "
          data-icon="eye-slash"
          data-prefix="far"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 640 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M634 471L36 3.51A16 16 0 0 0 13.51 6l-10 12.49A16 16 0 0 0 6 41l598 467.49a16 16 0 0 0 22.49-2.49l10-12.49A16 16 0 0 0 634 471zM296.79 146.47l134.79 105.38C429.36 191.91 380.48 144 320 144a112.26 112.26 0 0 0-23.21 2.47zm46.42 219.07L208.42 260.16C210.65 320.09 259.53 368 320 368a113 113 0 0 0 23.21-2.46zM320 112c98.65 0 189.09 55 237.93 144a285.53 285.53 0 0 1-44 60.2l37.74 29.5a333.7 333.7 0 0 0 52.9-75.11 32.35 32.35 0 0 0 0-29.19C550.29 135.59 442.93 64 320 64c-36.7 0-71.71 7-104.63 18.81l46.41 36.29c18.94-4.3 38.34-7.1 58.22-7.1zm0 288c-98.65 0-189.08-55-237.93-144a285.47 285.47 0 0 1 44.05-60.19l-37.74-29.5a333.6 333.6 0 0 0-52.89 75.1 32.35 32.35 0 0 0 0 29.19C89.72 376.41 197.08 448 320 448c36.7 0 71.71-7.05 104.63-18.81l-46.41-36.28C359.28 397.2 339.89 400 320 400z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </span>
    </div>
    <div
      className="registration-form__column registration-form__column--wide"
    >
      <div
        className="checkbox-wrapper"
      >
        
        <input
          checked={false}
          className="checkbox-input "
          id="receive-emails"
          name="acceptReceiveEmails"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          className="checkbox-label"
          htmlFor="receive-emails"
        >
          Yes, I want to receive Lottery Display emails
        </label>
      </div>
      <div
        className="checkbox-wrapper"
      >
        
        <input
          checked={false}
          className="checkbox-input "
          id="accept-terms"
          name="acceptTerms"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          className="checkbox-label"
          htmlFor="accept-terms"
        >
          I agree to the all
        </label>
        <p>
          <a
            className="registration-form__link"
            href="#"
          >
            Â Term,
             
          </a>
          <a
            className="registration-form__link"
            href="#"
          >
            Privacy Policy
             
          </a>
          and
           
          <a
            className="registration-form__link"
            href="#"
          >
            Fees
          </a>
        </p>
      </div>
    </div>
    <button
      className="registration-form__submit"
      type="submit"
    >
      Create Account
    </button>
    <div
      className="registration-form__column registration-form__column--wide"
    >
      <p>
        Already have an account?
         
        <a
          className="registration-form__link"
          href="#"
        >
          Log In
        </a>
      </p>
    </div>
  </form>
</div>
`;
